lib/barcelona/plugins/ossec_client_plugin.rb

```diff
@@ -0,0 +1,56 @@
+# OssecClient plugin
+# This plugin adds a wazuh agent which connects to existing ossec manager
+# Usage: bcn district put-plugin discrict1 ossec_client -a server_hostname=ossec-manager.local 
+# To connect a ossec manager in other VPC, you need to ...
+# - Add VPC to Route53 Hosted Zone
+# - Create a peering connection
+# - Add an entry for peering to both Container VPC and ossec manager VPC
+# - Allow access from Container VPC to ossec manager VPC
+
+module Barcelona
+  module Plugins
+    class OssecClientPlugin < Base
+      def on_container_instance_user_data(_instance, user_data)
+        user_data.extend OssecClientUserData
```

Comment:
Why do you extend `user_data` here while in the itamae plugin it's not? I prefer to not extend the user_data object because the same object is used in the core so extending the object might cause a problem in the core
(I don't think there's any problem right now but maybe in the future we may add something that depends on modules that `user_data` extends or its class includes.)

Reply:
I agree that consistency is important but I prefer separating UserData and plugin body by modules. So I'd like to re-write Itamae plugin in the same way as this one.

I think we should write new features as plugins from now on because we have important production districts and we don't want to break them on testing new features on staging district. Plugins can be enabled districts by districts

So we need a pattern for implementing coming plugins which may be more complicated. I'd like to recommend using modules and `extend` for following reasons.

1. module name and method name declares WHAT  while hiding HOW or implementation in method body
2. UserData should hide internals from its clients
3. some modules for UserData can be shared among many plugins
4. Later, we may need a callback with which Barcelona core calls plugin code at certain trigger, extending a module can be used naturally in such a case

I think UserData is an important object for plugins and the best way is encapsulate its internals by writing a class for it with a clear interface. But it will be a too big update and this pattern is a practical compromise

Reply:
I see, then how about defining a module with a class method like this?

```ruby
  module OssecClientUserData
    def self.configure_ossec_client(user_data, server_hostname)
      # ...
    end
  end

  # And then call the method
  OssecClientUserData.configure_ossec_client(user_data, attributes["server_hostname"])
```

This way we don't need to make the user_data object dirty.
The only thing I'm worried about is method name conflicts. As long as we know method name is not going to conflict I think it's fine.
I know the methond name `configure_ossec_client` will be very rarely conflict, but if this pattern becomes a rule it's possible that somebody define a module with a common name like `configure`. This kind of bug can be very difficult to detect in development time because the issue only occurs with a specific plugin combination and order.